java-checks/src/main/java/org/sonar/java/checks/StandardFunctionalInterfaceCheck.java

```diff
@@ -0,0 +1,299 @@
+    MethodTree declaration = functionalMethod.declaration();
+    if (!functionalMethod.thrownTypes().isEmpty() || (declaration != null && !declaration.typeParameters().isEmpty())) {
+      return Optional.empty();
+    }
+    return STD_INTERFACE_BY_PARAMETER_COUNT.getOrDefault(functionalMethod.parameterTypes().size(), Collections.emptyList()).stream()
+      .map(standardInterface -> standardInterface.matchingSpecialization(functionalMethod))
+      .filter(specialization -> specialization != null)
+      .findFirst();
+  }
+
+  private static boolean hasNoExtension(ClassTree tree) {
+    return tree.superInterfaces().isEmpty();
+  }
+
+  private static boolean isFunctionalInterface(ClassTree tree) {
+    return tree.symbol().metadata().isAnnotatedWith("java.lang.FunctionalInterface");
+  }
+
+  private static boolean isNonStandardFunctionalInterface(ClassTree tree) {
+    return !STD_INTERFACE_NAMES.contains(tree.symbol().type().fullyQualifiedName());
+  }
+
+  private static boolean isNotObjectMethod(MethodSymbol method) {
+    MethodTree declaration = method.declaration();
+    return declaration == null || !OBJECT_METHODS.anyMatch(declaration);
+  }
+
+  private static boolean isNonStaticNonDefaultMethod(MethodSymbol methodSymbol) {
+    return !methodSymbol.isStatic() && !isDefault(methodSymbol);
+  }
+
+  private static boolean isDefault(MethodSymbol methodSymbol) {
+    return (((JavaSymbol.MethodJavaSymbol) methodSymbol).flags() & Flags.DEFAULT) != 0;
+  }
+
+  private static MethodMatcher methodMatcherWithName(String name, String... parameters) {
+    MethodMatcher methodMatcher = MethodMatcher.create().typeDefinition(TypeCriteria.anyType()).name(name);
+    for (String parameter : parameters) {
+      methodMatcher = methodMatcher.addParameter(parameter);
+    }
+    return methodMatcher;
+  }
+
+  private static class FunctionalInterface {
+
+    private final String name;
+    private final List<String> genericTypes;
+    private final String returnType;
+    private final List<String> parameters;
+
+    private FunctionalInterface(String name, String returnType, String... parameters) {
+      int genericStart = name.indexOf('<');
+      if (genericStart != -1) {
+        this.name = name.substring(0, genericStart);
+        this.genericTypes = Arrays.asList(name.substring(genericStart + 1, name.length() - 1).split(","));
+      } else {
+        this.name = name;
+        this.genericTypes = Collections.emptyList();
+      }
+      this.returnType = returnType;
+      this.parameters = Arrays.asList(parameters);
+    }
+
+    private String getName() {
+      return name;
+    }
+
+    private int getGenericTypeCount() {
+      return genericTypes.size();
+    }
+
+    private int getParameterCount() {
+      return parameters.size();
+    }
+
+    @CheckForNull
+    private String matchingSpecialization(MethodSymbol method) {
+      List<Type> methodParameters = method.parameterTypes();
```

Comment:
This variable should be moved closer to its first usage. (to have first to deal with return types then params)